### PR TITLE
[ISSUE-28] Concat/ConcatWs support BinaryString as input parameter

### LINKED ISSUE #28: Concat/ConcatWs support BinaryString as input parameter
Currently GeaFlow uses BinaryString as its implementation for string, which provides better performance. However, at present, most of the built-in String Function uses string as the input parameter, and a direct conversion from BinaryString to String is required in the calculation phase, which has performance overhead. Please Implement the Concat/ConcatWs function with BinaryString as an input.

### PR DESCRIPTION
**Pull request by liuxiaocs7**
### What changes were proposed in this pull request?

close #28

### How was this PR tested?
- [x] Tests have Added for the changes
- [ ] Production environment verified

### CODE REVIEW

**Comment by liuxiaocs7** [comment]
Hi, community, some questions about this issue.

1. In Spark UTF8String, `BYTE_ARRAY_OFFSET` is used when calculating the offset, `BYTE_ARRAY_OFFSET` exists in GeaFlow too but the init value is 16 which cause error?

https://github.com/apache/spark/blob/a8ea35f7c2fbd80c3902707ecf224d50800ba622/common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java#L949-L953

2. Regarding the handling of `null` by `concat` and `concat_ws`, does it need to be consistent with the previous logic?

https://github.com/TuGraph-family/tugraph-analytics/blob/6ba974752ca31ccf40e54b2cd37437274febb46a/docs/docs-cn/application-development/dsl/build-in/string.md?plain=1#L92-L94

**Comment by Leomrlin** [comment]
> Hi, community, some questions about this issue.
> 
> 1. In Spark UTF8String, `BYTE_ARRAY_OFFSET` is used when calculating the offset, `BYTE_ARRAY_OFFSET` exists in GeaFlow too but the init value is 16 which cause error?
> 
> https://github.com/apache/spark/blob/a8ea35f7c2fbd80c3902707ecf224d50800ba622/common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java#L949-L953
> 
> 2. Regarding the handling of `null` by `concat` and `concat_ws`, does it need to be consistent with the previous logic?
> 
> https://github.com/TuGraph-family/tugraph-analytics/blob/6ba974752ca31ccf40e54b2cd37437274febb46a/docs/docs-cn/application-development/dsl/build-in/string.md?plain=1#L92-L94

hi, [liuxiaocs7](https://github.com/liuxiaocs7)

File: geaflow/geaflow-dsl/geaflow-dsl-plan/src/test/java/com/antgroup/geaflow/dsl/udf/string/UDFStringTest.java [hunk 1]
```diff
@@ -69,10 +70,48 @@ public void testBase64() {
     @Test
     public void testConcat() {
         String string = "ant group";
+        BinaryString binaryString = BinaryString.fromString(string);
         Concat test = new Concat();
         test.open(null);
-        assertEquals(test.eval(string, string, string),
+        assertEquals(test.eval(string, string, string), "ant groupant groupant group");
```
Comment (Leomrlin) [inline geaflow/geaflow-dsl/geaflow-dsl-plan/src/test/java/com/antgroup/geaflow/dsl/udf/string/UDFStringTest.java:76]: Can we add a few tests related to null here？

File: geaflow/geaflow-dsl/geaflow-dsl-plan/src/test/java/com/antgroup/geaflow/dsl/udf/string/UDFStringTest.java [hunk 2]
```diff
@@ -69,10 +70,48 @@ public void testBase64() {
...
+            BinaryString.fromString("蚂蚁3")), "蚂蚁1蚂蚁2蚂蚁3");
+    }
+
+    @Test
+    public void testConcatWS() {
+        String string = "ant group";
+        ConcatWS test = new ConcatWS();
+        BinaryString binaryString = BinaryString.fromString(string);
+        test.open(null);
+        assertEquals(test.eval(",", string, string, string),
```
Comment (Leomrlin) [inline geaflow/geaflow-dsl/geaflow-dsl-plan/src/test/java/com/antgroup/geaflow/dsl/udf/string/UDFStringTest.java:98]: Same as above, can we add a few tests related to null?

File: geaflow/geaflow-dsl/geaflow-dsl-plan/src/test/java/com/antgroup/geaflow/dsl/udf/string/UDFStringTest.java [hunk 3]
```diff
@@ -69,10 +70,48 @@ public void testBase64() {
...
+        assertEquals(test.eval("-", string, string, string),
+            "ant group-ant group-ant group");
+        assertEquals(test.eval("***", string, string, string),
+            "ant group***ant group***ant group");
+        assertEquals(test.eval("\t", string, string, string),
+            "ant group\tant group\tant group");
+        assertEquals(test.eval("\n", string, string, string),
+            "ant group\nant group\nant group");
+
+        assertEquals(
```
Comment (Leomrlin) [inline geaflow/geaflow-dsl/geaflow-dsl-plan/src/test/java/com/antgroup/geaflow/dsl/udf/string/UDFStringTest.java:113]: It's necessary to add a method that takes a separator as a normal String type？

**Review by Leomrlin — APPROVED** [review]
Thank you for implementing the concat and concatWs methods for binary. It is great that the code is consistent with the existing UDFs in GeaFlow! Would it be possible to optimize a few minor details?

**Comment by QiZhang1997** [comment]
> 



> Hi, community, some questions about this issue.
> 
> 1. In Spark UTF8String, `BYTE_ARRAY_OFFSET` is used when calculating the offset, `BYTE_ARRAY_OFFSET` exists in GeaFlow too but the init value is 16 which cause error?
> 
> https://github.com/apache/spark/blob/a8ea35f7c2fbd80c3902707ecf224d50800ba622/common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java#L949-L953
> 
> 2. Regarding the handling of `null` by `concat` and `concat_ws`, does it need to be consistent with the previous logic?
> 
> https://github.com/TuGraph-family/tugraph-analytics/blob/6ba974752ca31ccf40e54b2cd37437274febb46a/docs/docs-cn/application-development/dsl/build-in/string.md?plain=1#L92-L94

In question 1, the init value of `BYTE_ARRAY_OFFSET` is not a constant value 16. It is initialized in BinaryOperations class.

File: geaflow/geaflow-dsl/geaflow-dsl-plan/src/test/java/com/antgroup/geaflow/dsl/udf/string/UDFStringTest.java (hunk 3 above)
Comment (liuxiaocs7) [reply geaflow/geaflow-dsl/geaflow-dsl-plan/src/test/java/com/antgroup/geaflow/dsl/udf/string/UDFStringTest.java:113]: removed~

File: geaflow/geaflow-dsl/geaflow-dsl-plan/src/test/java/com/antgroup/geaflow/dsl/udf/string/UDFStringTest.java (hunk 1 above)
Comment (liuxiaocs7) [reply geaflow/geaflow-dsl/geaflow-dsl-plan/src/test/java/com/antgroup/geaflow/dsl/udf/string/UDFStringTest.java:76]: done!

**Comment by liuxiaocs7** [comment]
Hi, @Leomrlin, i have unified the logic and added some IT for `null`.

About this test case

```
concat(null) = null;
```

when `args` is `null`, when args is null, will there be NPE according to the code?

```java
public String eval(String... args) {
    StringBuilder sb = new StringBuilder();
    for (String arg : args) {
        if (arg != null) {
            sb.append(arg);
        }
    }
    return sb.toString();
}
```

**Comment by liuxiaocs7** [comment]
> > 
> 
> > Hi, community, some questions about this issue.
> > 
> > 1. In Spark UTF8String, `BYTE_ARRAY_OFFSET` is used when calculating the offset, `BYTE_ARRAY_OFFSET` exists in GeaFlow too but the init value is 16 which cause error?
> > 
> > [apache/spark@`a8ea35f`/common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java#L949-L953](https://github.com/apache/spark/blob/a8ea35f7c2fbd80c3902707ecf224d50800ba622/common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java#L949-L953)
> > 
> > 2. Regarding the handling of `null` by `concat` and `concat_ws`, does it need to be consistent with the previous logic?
> > 
> > https://github.com/TuGraph-family/tugraph-analytics/blob/6ba974752ca31ccf40e54b2cd37437274febb46a/docs/docs-cn/application-development/dsl/build-in/string.md?plain=1#L92-L94
> 
> In question 1, the init value of `BYTE_ARRAY_OFFSET` is not a constant value 16. It is initialized in BinaryOperations class.

Thank you for your timely feedback, I will take a deeper look at this part!

**Comment by Leomrlin** [comment]
> > > 
> > 
> > 
> > > Hi, community, some questions about this issue.
> > > 
> > > 1. In Spark UTF8String, `BYTE_ARRAY_OFFSET` is used when calculating the offset, `BYTE_ARRAY_OFFSET` exists in GeaFlow too but the init value is 16 which cause error?
> > > 
> > > [apache/spark@`a8ea35f`/common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java#L949-L953](https://github.com/apache/spark/blob/a8ea35f7c2fbd80c3902707ecf224d50800ba622/common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java#L949-L953)
> > > 
> > > 2. Regarding the handling of `null` by `concat` and `concat_ws`, does it need to be consistent with the previous logic?
> > > 
> > > https://github.com/TuGraph-family/tugraph-analytics/blob/6ba974752ca31ccf40e54b2cd37437274febb46a/docs/docs-cn/application-development/dsl/build-in/string.md?plain=1#L92-L94
> > 
> > 
> > In question 1, the init value of `BYTE_ARRAY_OFFSET` is not a constant value 16. It is initialized in BinaryOperations class.
> 
> Thank you for your timely feedback, I will take a deeper look at this part!

Oh Yes, I experimented with the existing concat and passing a null parameter causes an NPE. Can you add a null pointer check to fix this bug? Thanks

**Comment by liuxiaocs7** [comment]
> > Hi, @Leomrlin, i have unified the logic and added some IT for `null`.
> > About this test case
> > ```
> > concat(null) = null;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > when `args` is `null`, when args is null, will there be NPE according to the code?
> > ```java
> > public String eval(String... args) {
> >     StringBuilder sb = new StringBuilder();
> >     for (String arg : args) {
> >         if (arg != null) {
> >             sb.append(arg);
> >         }
> >     }
> >     return sb.toString();
> > }
> > ```
> 
> Oh Yes, I experimented with the existing concat and passing a null parameter causes an NPE. Can you add a null check to fix this bug? Thanks

sure, i want to have a try, BTW `concat(null, null, null) = null`  is it the expected result?

in this condition, `args` is a string array which contains there elements and every element in `null`.

**Comment by Leomrlin** [comment]
> concat('1',null,'2') = '12' 
>  concat('1','2',null) = '12' 
>  concat(null) = null;

The call of concat(null, null, null) will result in an empty string, which is not a problem in the existing code. However, a single null parameter will cause an NPE, which is not consistent with the expectation in the documentation. concat(null) does not seem like a reasonable call. Generally, concat should have two or more parameters, and if the only parameter is null, it should return null, consistent with the documentation: [https://tugraph-analytics.readthedocs.io/en/latest/docs-en/application-development/dsl/build-in/string](https://tugraph-analytics.readthedocs.io/en/latest/docs-en/application-development/dsl/build-in/string/%EF%BC%9F)

**Comment by liuxiaocs7** [comment]
> > sure, i want to have a try, BTW `concat(null, null, null) = null` is it the expected result?
> > in this condition, `args` is a string array which contains there elements and every element in `null`.
> 
> The call of concat(null, null, null) will result in an empty string, which is not a problem in the existing code. However, a single null parameter will cause an NPE, which is not consistent with the expectation in the documentation. concat(null) does not seem like a reasonable call. Generally, concat should have two or more parameters, and if the only parameter is null, it should return null, consistent with the documentation: [tugraph-analytics.readthedocs.io/en/latest/docs-en/application-development/dsl/build-in/string](https://tugraph-analytics.readthedocs.io/en/latest/docs-en/application-development/dsl/build-in/string)

got it, make sense, thanks!

File: geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/ConcatWS.java [hunk 4]
```diff
@@ -14,14 +14,20 @@
...
 
 @Description(name = "concat_ws", description = "Concat strings to one by the separator string.")
 public class ConcatWS extends UDF {
 
     public String eval(String separator, String... args) {
-        return StringUtils.join(separator, args);
+        return StringUtils.join(args, separator);
+    }
+
+    public String eval(BinaryString separator, BinaryString... args) {
```
Comment (Leomrlin) [inline geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/ConcatWS.java]: ConcatWS needs an eval method that takes the parameter separator as a regular String type instead of BinaryString. Otherwise, calls like concatWS("-", test1, test2) in DSL may not find the method corresponding to the type.

**Review by Leomrlin — CHANGES_REQUESTED** [review]

File: geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/ConcatWS.java (hunk 4 above)
Comment (liuxiaocs7) [reply geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/ConcatWS.java]: updated

**Review by Leomrlin — APPROVED** [review]
LGTM

File: geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/Concat.java [hunk 5]
```diff
@@ -29,4 +34,11 @@ public String eval(String... args) {
         }
         return sb.toString();
     }
+
+    public String eval(BinaryString... args) {
```
Comment (pengzhiwei2018) [inline geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/Concat.java]: Please use BinaryString as the return type.

File: geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/ConcatWS.java [hunk 6]
```diff
@@ -14,14 +14,20 @@
...
 
 @Description(name = "concat_ws", description = "Concat strings to one by the separator string.")
 public class ConcatWS extends UDF {
 
     public String eval(String separator, String... args) {
-        return StringUtils.join(separator, args);
+        return StringUtils.join(args, separator);
+    }
+
+    public String eval(String separator, BinaryString... args) {
```
Comment (pengzhiwei2018) [inline geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/ConcatWS.java]: Use BinaryString as the return type

File: geaflow/geaflow-common/src/main/java/com/antgroup/geaflow/common/binary/BinaryString.java [hunk 7]
```diff
@@ -226,6 +229,67 @@ public boolean contains(final BinaryString substring) {
...
+                continue;
+            }
+            int len = input.numBytes;
+            copyMemory(input.binaryObject, input.offset, result, offset, len);
+            offset += len;
+        }
+        return fromBytes(result);
+    }
+
+    public static BinaryString concatWs(String separator, BinaryString... inputs) {
```
Comment (pengzhiwei2018) [inline geaflow/geaflow-common/src/main/java/com/antgroup/geaflow/common/binary/BinaryString.java]: The type for `separator`  should also be `BinaryString`

Comment (liuxiaocs7) [reply geaflow/geaflow-common/src/main/java/com/antgroup/geaflow/common/binary/BinaryString.java]: done!

File: geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/ConcatWS.java (hunk 6 above)
Comment (pengzhiwei2018) [reply geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/ConcatWS.java]: Hi @liuxiaocs7 can you take some time to resolve the rest comment?

Comment (liuxiaocs7) [reply geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/ConcatWS.java]: Hi, @pengzhiwei2018, i have resolved comments, could you take a look when you are free.

File: geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/ConcatWS.java [hunk 8]
```diff
@@ -14,14 +14,22 @@
...
 
 @Description(name = "concat_ws", description = "Concat strings to one by the separator string.")
 public class ConcatWS extends UDF {
 
     public String eval(String separator, String... args) {
-        return StringUtils.join(separator, args);
+        return StringUtils.join(args, separator);
+    }
+
+    public BinaryString eval(String separator, BinaryString... args) {
```
Comment (pengzhiwei2018) [inline geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/ConcatWS.java:30]: Use BinaryString as the input type for `separator`

File: geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/ConcatWS.java [hunk 9]
```diff
@@ -14,14 +14,22 @@
...
+        BinaryString sep = separator == null ? BinaryString.EMPTY_UTF8 :
+                           BinaryString.fromString(separator);
+        return Objects.requireNonNull(BinaryString.concatWs(sep, args));
```
Comment (pengzhiwei2018) [inline geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/ConcatWS.java]: remove the `NonNull` check for the result.

File: geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/ConcatWS.java (hunk 8 above)
Comment (liuxiaocs7) [reply geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/ConcatWS.java:30]: According to this comment https://github.com/TuGraph-family/tugraph-analytics/pull/45#discussion_r1241105281, change request to use String as the input type for separator?

Comment (Leomrlin) [reply geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/table/string/ConcatWS.java:30]: > According to this comment [#45 (comment)](https://github.com/TuGraph-family/tugraph-analytics/pull/45#discussion_r1241105281), change request to use String as the input type for separator?

Both the eval methods with String and BinaryString as separator need to be added

**Review by pengzhiwei2018 — APPROVED** [review]
LGTM! Thanks @liuxiaocs7 for the contribution.
